### PR TITLE
perf(db_handler): use LRU eviction for backend prepared statements

### PR DESCRIPTION
**Pull request by MorganaFuture**
## What kind of change does this PR introduce?

Performance improvement.

## What is the current behavior?

When the per-backend prepared statement set hits the 200-statement limit, eviction picks 40 entries at random (`Enum.take_random`) and sends `Close` for them. Hot statements get evicted with the same probability as cold ones,
so under churn we can drop a frequently used statement and pay for a fresh `Parse` round-trip on the next `Bind`.    Relates to #69 / #239.

## What is the new behavior?
   
Replaces the `MapSet`-backed set with a small LRU storage (`Supavisor.Protocol.PreparedStatements.BackendStorage`). Recency is bumped on `Parse` and on `Bind` for statements already known to the backend. Eviction now drops the least recently used statements first, so hot ones stay around.

No default behavior change still gated by the `named_prepared_statements` feature flag. Telemetry (`prepared_statements_evicted`) is unchanged. Covered by unit tests (basic API + edge cases) and a `StreamData` property test that checks the LRU ordering invariant across arbitrary `put`/`touch`/`delete` sequences.

### CODE REVIEW

**Review by v0idpwn — COMMENTED** [review]
Thanks for the PR, @MorganaFuture, it looks great conceptually and the implementation is sound! I wanted to implement something like this but didn't get around to do it.

I'd like to suggest an architectural change, though: I'd like to make the db handler prepared statement storage backend pluggable. Why: we haven't really adopted this yet in production, and it would be nice to have both implementations at the same time so we can compare them in different scenarios. We could select it based on a feature flag, and eventually settle in a default. This also sets us up for a future in case we want to try a third storage.

WDYT?

**Comment by MorganaFuture** [comment]
> Thanks for the PR, @MorganaFuture, it looks great conceptually and the implementation is sound! I wanted to implement something like this but didn't get around to do it.
> 
> I'd like to suggest an architectural change, though: I'd like to make the db handler prepared statement storage backend pluggable. Why: we haven't really adopted this yet in production, and it would be nice to have both implementations at the same time so we can compare them in different scenarios. We could select it based on a feature flag, and eventually settle in a default. This also sets us up for a future in case we want to try a third storage.
> 
> WDYT?

Great suggestion. BackendStorage is now a behaviour with two impls LRU and Random. Random is default.
Also could you please take a look at this pr [978](https://github.com/supabase/supavisor/pull/978). I am little annoyed by the failing CI/CD. Let's fix it.

File: lib/supavisor/protocol/prepared_statements/backend_storage.ex [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  @type t() :: struct()
+  @type name() :: PreparedStatements.statement_name()
+
+  @callback new() :: t()
+  @callback size(t()) :: non_neg_integer()
+  @callback member?(t(), name()) :: boolean()
+  @callback put(t(), name()) :: t()
+  @callback touch(t(), name()) :: t()
+  @callback delete(t(), name()) :: t()
+  @callback pop_oldest(t(), pos_integer()) :: {[name()], t()}
```
Comment (v0idpwn) [inline lib/supavisor/protocol/prepared_statements/backend_storage.ex]: Nitpick: for `Random`, these aren't the oldest, so it should probably be called `pop_n`

**Review by v0idpwn — APPROVED** [review]
Looks great, just one small note!

**Comment by v0idpwn** [comment]
Regarding #978, we will do the bump separately as it requires appups to ship these version upgrades with zero downtime in production. These appups can be a bit nasty to write and test (and potentially not possible), so we will take our time to explore the version bumping in the next couple days.

File: lib/supavisor/protocol/prepared_statements/backend_storage.ex (hunk 1 above)
Comment (v0idpwn) [reply lib/supavisor/protocol/prepared_statements/backend_storage.ex]: Or maybe `evict`

Comment (v0idpwn) [reply lib/supavisor/protocol/prepared_statements/backend_storage.ex]: gentle nudge @MorganaFuture, then I can merge!

Comment (MorganaFuture) [reply lib/supavisor/protocol/prepared_statements/backend_storage.ex]: @v0idpwn Done

**Comment by MorganaFuture** [comment]
@v0idpwn  So let's merge?

**Comment by v0idpwn** [comment]
Just wanted to thank you for this one @MorganaFuture, sorry for how long it took to merge! 💚
